### PR TITLE
Rate limiter

### PR DESCRIPTION
**Pull request by samgermain**
@kroitor @frosty00 @DoctorSlimm 

So I'm trying to edit the rate limiter so that the max amount of your available limit is used before your requests start being slowed down, this would make 60 1 weight requests all executed asynchronously at the same time if 60 is not above the rate limit.

Besides the code in throttle.js and Exchange.initRestRateLimiter, I'm not sure what other code would need to be edited, I'm not sure exactly what should be done.

### CODE REVIEW

File: js/base/functions/throttle.js [hunk 1]
```diff
@@ -15,61 +15,63 @@ class Throttle {
...
+        let lastTimestamp = now ();
+        while (this.running) { // loops through method calls in the queue, executing them if tokens available, and waiting if tokens not available
+            const { resolver, cost } = this.queue[0];
+            if (this.config['tokens'] >= 0) { // if rate limit hasn't been reached
+                this.config['tokens'] -= cost;
+                resolver (); // method for api endpoint
+                this.queue.shift ();
                 // contextswitch
-                await Promise.resolve ()
+                await Promise.resolve (); // what does this do?
```
Comment (samgermain) [inline js/base/functions/throttle.js:37]: I wasn't sure what this line was for

File: js/base/functions/throttle.js [hunk 2]
```diff
@@ -15,61 +15,67 @@ class Throttle {
...
-        this.queue = []
-        this.running = false
+        };
+        Object.assign (this.config, config);
+        this.queue = [];  // requests to be sent
+        this.running = false;
+    }
+
+    async resolve (resolver) {
+        resolver (); // method for api endpoint
```
Comment (samgermain) [inline js/base/functions/throttle.js:25]: so this would be the big change that would get the available limit to fill up faster. Would it work how I think?

**Comment by kroitor** [comment]
In short, we should not really change the existing code, we should probably add a new rate limiting class alongside, and maybe rearrange those classes into a set of rate limiting policies. At this point it breaks the build. I will think a bit more and will try to suggest a flexible design for those purpose.

**Comment by samgermain** [comment]
> In short, we should not really change the existing code, we should probably add a new rate limiting class alongside, and maybe rearrange those classes into a set of rate limiting policies. At this point it breaks the build. I will think a bit more and will try to suggest a flexible design for those purpose.

I can do that, but at this point It's more important that I understand if my changes will work, once I have a clear picture of what needs to be done I can split it up

**Comment by frosty00** [comment]
i'm going to close this because the edits aren't really that clean

> So I'm trying to edit the rate limiter so that the max amount of your available limit is used before your requests start being slowed down, this would make 60 1 weight requests all executed asynchronously at the same time if 60 is not above the rate limit.

To do this we should just increase the size of the capacity of the tokenBucket. Also let's add the edits to all the langs for consistency.

**Pull request closed without merging**